### DISCUSSION TITLE
Postgres Language Server: JS + TS Template Literal Support

### DISCUSSION THREAD

**Discussion by karlhorky** [supabase/supabase#16286] [Feature Requests]
Hi there 👋 First of all, thanks for all of the effort that Supabase is putting into the PostgreSQL community, including tooling, really amazing 🙌 

I recently encountered Postgres Language Server via [the recent tweet](https://twitter.com/karlhorky/status/1688489997240881152) from @kiwicopple and it looks amazing!

I have been chasing the "holy grail" of tooling for SQL template literals in JavaScript + TypeScript for a while now:

- syntax highlighting (works, with the VS Code extension [SQL tagged template literals (syntax only)](https://marketplace.visualstudio.com/items?itemName=frigus02.vscode-sql-tagged-template-literals-syntax-only))
- linting (works, using [SafeQL](https://github.com/ts-safeql/safeql/))
- formatting, some options
  - [`prettier-plugin-embed` and `prettier-plugin-sql`](https://github.com/Sec-ant/prettier-plugin-embed/issues/20) - uses [`sql-formatter`](https://github.com/sql-formatter-org/sql-formatter)
  - [kind of works, with `eslint-plugin-sql` and `eslint-plugin-unicorn`](https://github.com/gajus/eslint-plugin-sql/pull/33#issuecomment-1666585459), but very opinionated / not so configurable - uses [`pg-formatter`](https://github.com/gajus/pg-formatter)
  - [`@potygen/prettier-plugin-pgsql`](https://github.com/ivank/potygen/blob/main/packages/prettier-plugin-pgsql/README.md) by @ivank - uses custom Prettier AST formatter

Wasn't even looking for inline documentation yet, but that's a great addition too!

It seems that the plan will be for Postgres Language Server to support all of these use cases, which is amazing!

I would love to have everything handled by one highly-battle-tested tool, which could maybe also deal with some of the shortcomings from the tools listed above.

However, for Postgres Language Server to work for any of our projects, it would need a feature which I'm not sure it currently has (can't tell, from a quick glance at the docs and the code):

- Support for tagged template literals in JavaScript and TypeScript

There is precedent for SQL tagged template literals in JS+TS, some of the more popular ones being:

- [Postgres.js](https://github.com/porsager/postgres)
- [`@vercel/postgres`](https://vercel.com/docs/storage/vercel-postgres/sdk)

Eg. I would love for all of the features of Postgres Language Server to be available in the following code block:

```ts
import { createClient } from '@vercel/postgres';
 
async function queryPosts() {
  const client = createClient();
  await client.connect();
 
  try {
    const likes = 100;
    const { rows, fields } =
      await client.sql`SELECT * FROM posts WHERE likes > ${likes};`;
  } finally {
    await client.end();
  }
}
```

Or this code block:

```ts
// db.js
import postgres from 'postgres'

const sql = postgres({ /* options */ }) // will use psql environment variables

async function getUsersOver(age) {
  const users = await sql`
    select
      name,
      age
    from users
    where age > ${ age }
  `
  // users = Result [{ name: "Walter", age: 80 }, { name: 'Murray', age: 68 }, ...]
  return users
}


async function insertUser({ name, age }) {
  const users = await sql`
    insert into users
      (name, age)
    values
      (${ name }, ${ age })
    returning name, age
  `
  // users = Result [{ name: "Murray", age: 68 }]
  return users
}
```

**Comment by kiwicopple** [comment]
sounds great @karlhorky - the support for this will largely depend on the code editor you're using. 

For example, here are the docs for VSCode: https://code.visualstudio.com/api/references/vscode-api#languages. It's up to the code editor to "guess" which language is defined, and then choose the appropriate Language Editor. So this should be solved once we make progress on the language server.

FYI - the language server is [here](https://github.com/supabase/postgres_lsp) in case you want to discuss in a more relevant location :)

**Reply by karlhorky** [reply]
sure, I'll reopen my issue (I closed because I thought it may not be a "bug"):

- https://github.com/supabase/postgres_lsp/issues/20

**Reply by karlhorky** [reply]
Since this probably is going to be a long-running issue, it was decided to make it into a discussion over in the repo here:

- https://github.com/supabase/postgres_lsp/discussions/25
